MSC4505: Push Rules for Live Location Sharing (#4505)
matrix-org/matrix-doc: proposals/4505-lls-push-rules.md

Reviewer: Implementation requirements:

* [ ] Client (using)
* [ ] Server (respecting)

UX for setting the push rules is not required. We have a number of push rules already which demonstrate the UX possibilities (including a flight console).

Reviewer: From PR description:

 * Synapse impl: https://github.com/element-hq/synapse/pull/19954
 * EXI impl: https://github.com/element-hq/element-x-ios/pull/5872
 * EXA impl: https://github.com/element-hq/element-x-android/pull/7203
 * EW impl: https://github.com/element-hq/element-web/pull/34247

    {"kind": "event_property_is", "key": "content.live", "value": true}
  ],
  "actions": [
    "notify"
  ]
}
```

Additionally, for unencrypted rooms, a new override rule is defined to explicitly suppress the `m.beacon`
location update events, inserted immediately after the `.m.rule.poll_response` override rule. 
This stops sending needless push for every LLS beacon update.

```jsonc
{
  "rule_id": ".m.rule.beacon",
  "default": true,
  "enabled": true,
  "conditions": [
    {"kind": "event_property_is", "key": "type", "value": "m\\.beacon"}
  ],
  "actions": []
}
```

## Potential issues

The `.m.rule.beacon` suppression cannot take effect in encrypted rooms, where location updates are
sent as `m.room.encrypted` events and therefore match the `.m.rule.encrypted` underride on the
server. Clients which re-evaluate push rules against the decrypted event (as the major
implementations do) will suppress the notification via `.m.rule.beacon` at that point; the cost of
the spurious push wakeups themselves is a pre-existing problem for all encrypted event types that
are uninteresting once decrypted, and is out of scope here.

> Re-sending `m.beacon_info` with `live: true` (e.g. extending a share's timeout) will notify again.
> This is arguably correct ("X is (still) sharing their live location"); clients may wish to
> de-duplicate notifications for a beacon they already know about.

Reviewer 2: Du-duplicating itself seems ok. The wording here doesn't sound fully convincing though. A notification for something that hasn't changed feels incorrect rather than correct to me. 🤔

Maybe we could just omit the correctness part, accept duplicates as a fact and RECOMMEND clients to handle them?

```suggestion
Re-sending `m.beacon_info` with `live: true` (e.g. extending a share's timeout) will notify again.
Clients SHOULD de-duplicate notifications for a beacon they already know about.
```

Reviewer 3: On the client side, I don't think we'd want to send another notification in this case. We do want to send the initial notification, but we won't put the duration in it. If the person sharing extends it, I don't think we'd notify at all. So this will end up being more code on the client side to handle this. Unless there's a clear use case for notifications in this scenario, I'd suggest we go the other way and eliminate this as the default.
